rust/connlib/tunnel/src/client.rs

```diff
@@ -847,10 +824,18 @@ impl ClientState {
             .chain(self.dns_mapping.left_values().copied().map(Into::into))
     }
 
-    fn get_cidr_resource_by_destination(&self, destination: IpAddr) -> Option<ResourceId> {
-        self.cidr_resources
+    fn get_resource_by_destination(&self, destination: IpAddr) -> Option<ResourceId> {
+        let maybe_cidr_resource_id = self
+            .cidr_resources
             .longest_match(destination)
-            .map(|(_, res)| res.id)
+            .map(|(_, r)| r.id);
+
+        let maybe_dns_resource_id = self
+            .dns_resources_internal_ips
+            .iter()
+            .find_map(|(r, i)| i.contains(&destination).then_some(r.id));
```

Comment:
Can this cause a performance regression? Might be worth the extra state keeping an additional table mapping ip addrs to peers?

Reply:
Yeah I thought about that too. I think using an `IpNetworkTable` would make sense here. More ergonomic to use as well.

Reply:
I don't think we need that, since we always assign `IpAddr`s for dns resources or you want to combine both CIDR and DNS on a single table?

Reply:
Ah, you mean a regular `HashMap`. Yeah that makes more sense. We'd always to an `exact_match` so it is redundant to use an `IpNetworkTable`.

Reply:
Hmm, it could be combined though! Like into a more general "routing table":

- CIDR resources go straight into that routing table
- Resolved DNS IPs also go into that table

Each packet goes through the path of:

- Lookup destination IP in routing table to find our `ResourceId`
- Lookup gateway based on resource ID
- Connect to gateway if aren't yet connected

Reply:
I've captured this idea in https://github.com/firezone/firezone/issues/5114.